Nil dereference panic in dialer

```
Panic at 2020-02-26T12:45:00+01:00
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x92d33e]

goroutine 74 [running]:
github.com/syncthing/syncthing/lib/dialer.(*dialerConn).Close(0xc000ee0060, 0xc00318c180, 0x110a780)
	<autogenerated>:1 +0x2e
github.com/syncthing/syncthing/lib/osutil.TCPPing(0x111d1e0, 0xc000454080, 0xc000b68ea8, 0x14, 0x0, 0x0, 0x0)
	/home/ckellermann/src/playground/syncthing/lib/osutil/ping.go:26 +0x1d2
github.com/syncthing/syncthing/lib/osutil.GetLatencyForURL(0x111d1e0, 0xc000454080, 0xc000b68ea0, 0xc7, 0x575ea59, 0x0, 0x0)
	/home/ckellermann/src/playground/syncthing/lib/osutil/ping.go:39 +0x94
github.com/syncthing/syncthing/lib/relay/client.relayAddressesOrder(0x111d1e0, 0xc000454080, 0xc000e66000, 0x10a, 0x200, 0xc000e66000, 0x100, 0x200)
	/home/ckellermann/src/playground/syncthing/lib/relay/client/dynamic.go:162 +0x47c
github.com/syncthing/syncthing/lib/relay/client.(*dynamicClient).serve(0xc000130690, 0x111d1e0, 0xc000454080, 0xc0000a3180, 0x7ff863e00008)
	/home/ckellermann/src/playground/syncthing/lib/relay/client/dynamic.go:79 +0x75f
github.com/syncthing/syncthing/lib/relay/client.newCommonClient.func1(0x111d1e0, 0xc000454080, 0x0, 0x0)
	/home/ckellermann/src/playground/syncthing/lib/relay/client/client.go:62 +0x81
github.com/syncthing/syncthing/lib/util.(*service).Serve(0xc0004612c0)
	/home/ckellermann/src/playground/syncthing/lib/util/utils.go:247 +0x142
created by github.com/syncthing/syncthing/lib/connections.(*relayListener).serve
	/home/ckellermann/src/playground/syncthing/lib/connections/relay_listen.go:56 +0x141
```

#6363 

Because it returns a non-nil conn on error that is nonetheless nil internally, and other code does the wrong thing.

--- a/lib/dialer/public.go
+++ b/lib/dialer/public.go
@@ -76,10 +76,10 @@ func dialContextWithFallback(ctx context.Context, fallback proxy.ContextDialer, 
 	if noFallback {
 		conn, err := dialer.DialContext(ctx, network, addr)
 		l.Debugf("Dialing no fallback result %s %s: %v %v", network, addr, conn, err)
-		conn = dialerConn{
-			conn, newDialerAddr(network, addr),
+		if err != nil {
+			return nil, err
 		}
-		return conn, err
+		return dialerConn{conn, newDialerAddr(network, addr)}, nil
 	}
 
 	ctx, cancel := context.WithCancel(ctx)
@@ -91,8 +91,8 @@ func dialContextWithFallback(ctx context.Context, fallback proxy.ContextDialer, 
 	go func() {
 		proxyConn, proxyErr = dialer.DialContext(ctx, network, addr)
 		l.Debugf("Dialing proxy result %s %s: %v %v", network, addr, proxyConn, proxyErr)
-		proxyConn = dialerConn{
-			proxyConn, newDialerAddr(network, addr),
+		if proxyErr != nil {
+			proxyConn = dialerConn{proxyConn, newDialerAddr(network, addr)}
 		}
 		close(proxyDone)
 	}()

--- a/lib/dialer/public.go
+++ b/lib/dialer/public.go
@@ -91,7 +91,7 @@ func dialContextWithFallback(ctx context.Context, fallback proxy.ContextDialer, 
 	go func() {
 		proxyConn, proxyErr = dialer.DialContext(ctx, network, addr)
 		l.Debugf("Dialing proxy result %s %s: %v %v", network, addr, proxyConn, proxyErr)
-		if proxyErr != nil {
+		if proxyErr == nil {
 			proxyConn = dialerConn{proxyConn, newDialerAddr(network, addr)}
 		}
 		close(proxyDone)

--- a/lib/osutil/ping.go
+++ b/lib/osutil/ping.go
@@ -22,7 +22,7 @@ func TCPPing(ctx context.Context, address string) (time.Duration, error) {
 	ctx, cancel := context.WithTimeout(ctx, time.Second)
 	defer cancel()
 	conn, err := dialer.DialContext(ctx, "tcp", address)
-	if conn != nil {
+	if err == nil {
 		conn.Close()
 	}
 	return time.Since(start), err

